review/searchTab.js

```diff
@@ -0,0 +1,108 @@
+
+SearchTab.prototype = {
+  constructor: SearchTab,
+  init() {
+    this.registerEvent();
+    this.renderSearchTab();
+  },
+  registerEvent() {
+    this.searchInput.addEventListener('input', this.handleInput.bind(this));
+    this.searchInput.addEventListener('keydown', this.handleKeydown.bind(this));
+  },
+  handleInput({ target: { value } }) {
+    this.timer = debounce(this.render.bind(this, value), 1000, this.timer);
+  },
+  handleKeydown({ keyCode }) {
+    if (keyCode === KEYCODE.UP) this.moveUpList();
+    else if (keyCode === KEYCODE.DOWN) this.moveDownList();
+  },
+  getRecommendHTML() {
+    const { SEARCH_TAB_LIST } = CLASS_LIST;
+    let firstList = '';
+    let secondList = '';
+    this.recommendData.forEach((data, idx) => {
+      if (idx < this.recommendData.length / 2) firstList += makeRecommendItem(idx + 1, data);
+      else secondList += makeRecommendItem(idx + 1, data);
+    });
+    const recommendHTML =
+      ul({ value: firstList, classes: [SEARCH_TAB_LIST] }) + ul({ value: secondList, classes: [SEARCH_TAB_LIST] });
+    return recommendHTML;
+  },
+  getAutoCompleteHTML() {
+    const { AUTOCOMPLETE_LIST } = CLASS_LIST;
+    const autoCompleteList = this.autoCompleteData.reduce((acc, autoData, idx) => {
+      if (idx === this.currentIdx)
+        return acc + makeAutoCompleteItem({ value: autoData, keyword: this.orginInput, isCurrentValue: true });
+      else return acc + makeAutoCompleteItem({ value: autoData, keyword: this.orginInput });
+    }, '');
+    const autoCompleteHTML = ul({ value: autoCompleteList, classes: [AUTOCOMPLETE_LIST] });
+    return autoCompleteHTML;
+  },
+  renderSearchTab() {
+    this.showTitle();
+    this.searchTab.innerHTML = this.getRecommendHTML();
+  },
+  renderAutoComplete() {
+    this.hiddenTitle();
+    this.searchTab.innerHTML = this.getAutoCompleteHTML();
+  },
+  async render(inputValue) {
+    this.initAutoCompleteIdx();
+    this.orginInput = inputValue;
+    await this.setAutoCompleteData(inputValue);
+    if (!this.autoCompleteData.length) this.renderSearchTab();
+    else this.renderAutoComplete();
+  },
+  async setAutoCompleteData(inputValue) {
+    const autoCompleteData = await getData(URL.autoComplete(inputValue));
+    this.autoCompleteData = autoCompleteParser(autoCompleteData);
+  },
+  showTitle() {
+    this.searchTabTitle.classList.remove(CLASS_LIST.HIDDEN);
+  },
+  hiddenTitle() {
+    this.searchTabTitle.classList.add(CLASS_LIST.HIDDEN);
+  },
+  moveUpList() {
+    if (this.currentIdx - 1 < 0) {
+      this.backUpInputValue();
+    } else {
+      this.currentIdx--;
+      this.searchInput.value = this.autoCompleteData[this.currentIdx];
+    }
+    this.renderAutoComplete();
+  },
+  moveDownList() {
+    if (this.currentIdx + 1 >= this.autoCompleteData.length) {
+      this.backUpInputValue();
+    } else {
+      this.currentIdx++;
+      this.searchInput.value = this.autoCompleteData[this.currentIdx];
+    }
+    this.renderAutoComplete();
+  },
```

Comment:
이 정도 중복은 그냥 둬도 크게 상관 없다고 생각 합니다. 하지만 공부하는 단계에서는 최대한으로 중복을 줄여보는 것도 좋을 것 같아요.

Reply:
> 이 정도 중복은 그냥 둬도 크게 상관 없다고 생각 합니다. 하지만 공부하는 단계에서는 최대한으로 중복을 줄여보는 것도 좋을 것 같아요.

저도 이 중복을 제거하고 싶었는데 마땅한 생각이 떠오르지 못해서 리팩토링을 진행하지 못했습니다..ㅜ 

조건문으로 up,down을 구분해서 하는 방법 밖에 떠오르지 않는데 괜찮을까요?

Reply:
중복되는 구조는 추상화 하고 값이나 조건이 달라지는 부분을 공통 메서드의 인자로 전달하는 방식으로 해볼 수 있겠죠 🙂 
```suggestion
  _moveList(needBackup, step) {
    if (needBackup) {
      this.backUpInputValue();
    } else {
      this.currentIdx += step;
      this.searchInput.value = this.autoCompleteData[this.currentIdx];
    }
    this.renderAutoComplete();
  },
  moveUpList() {
    _moveList(this.currentIdx - 1 < 0, -1);
  },
  moveDownList() {
    _moveList(this.currentIdx + 1 >= this.autoCompleteData.length, 1);
  },
```